### ISSUE TITLE
Slow performance

### ISSUE THREAD

**Issue by captkronos** [dnGrep/dnGrep#268]
Is there any way to increase the search speed of dnGrep? It is probably not a very typical use case but I have tried searching a folder of approximately 20TB of text files for a single text string and it takes dnGrep 19 minutes to complete. Whereas, a command line grep (Gnu Grep 2.24) only takes 4 minutes.

Thanks.

**Comment by JVimes** [comment]
I'm sure it could use a performance checkup. In the meantime, try unchecking this box and see if it helps:
![image](https://cloud.githubusercontent.com/assets/6148201/26374967/6db8a3e4-3fc4-11e7-9ebb-dd81368f72f3.png)

**Comment by doug24** [comment]
How many files are in the directory, and what is the range of file sizes?  Are they all text files?  Are there any zip files or other archive formats in there?

At my day job, we do have file shares with thousands of files in a single directory.  We've had much better performance iterating over the files using native Win32 APIs than with .NET APIs.  Doing something similar in dnGrep may be a good improvement.

**Comment by captkronos** [comment]
JVimes: Unchecking the verbose count didn't make any noticeable difference in the time.

doug24: There are approximately 100 files, all pure text (no archives), and all around the same size give or take a few hundred MB! It would be a great shame if the .NET APIs are so inefficient.

The test I am using for the timing only results in two hits so there isn't a large amount of GUI updating taking place.

**Comment by doug24** [comment]
I that case, it is probably not slow from iterating though lots of files.   A couple of things to try:

1) Change the Encoding selection to correct file encoding, and not the default "Auto detection".  Open Notepad++ will tell you the file encoding in its status bar.  This may not get you all the way there, but it could help.  Here's what I found:
Searching for a single match in 776MB text file went from 1 minute 20 seconds to 19 seconds when I pre-selected the encoding (UTF-8).
Searching for a single match in two text files (5GB and 1GB) went from 4:10 to 3:05 by pre-selecting the encoding (UTF-8).

2) I assume you are searching by regular expression: if you are, can you try searching for a text string instead, just to see if it makes a difference?  In my testing, using a regex was slightly slower, but not really significant.

**Comment by captkronos** [comment]
1. I changed the encoding to UTF-8, as per Notepad++, and the time went down to 15mins 15s. Which is a similar reduction to the 4:10 to 3:05 you saw. So a considerable decrease but still a long way from Grep.
2. I was already searching using the Text setting. As an experiment I tried Regex and the time went up to 20 mins (still using UTF-8).

**Comment by captkronos** [comment]
I've been playing around with a bit of vb.net and just in case it's of any use/interest, here are my findings.

### As a reference:
grep searchstring *.txt - takes 4:17 (4 minutes, 17 seconds)
grep -i searchstring *.txt - takes 4:17 (same as without -i)

### StreamReader with no encoding takes 5:36
```
Using sr As New StreamReader(filename)
	Dim aline As String
	While Not sr.EndOfStream
		aline = sr.ReadLine
		If aline.IndexOf(srchText) <> -1 Then
			Console.WriteLine(aline)
		End If
	End While
	sr.Close()
End Using
```

### StreamReader with encoding takes 6:06
```
Using sr As New StreamReader(filename, Encoding.Default)
	Dim aline As String
	While Not sr.EndOfStream
		aline = sr.ReadLine
		If aline.IndexOf(srchText) <> -1 Then
			Console.WriteLine(aline)
		End If
	End While
	sr.Close()
End Using
```


### StreamReader with no encoding and "ToLower" added (to be comparable with grep -i) takes 7:57
```
Using sr As New StreamReader(filename)
	Dim aline As String
	While Not sr.EndOfStream
		aline = sr.ReadLine.ToLower
		If aline.IndexOf(srchText) <> -1 Then
			Console.WriteLine(aline)
		End If
	End While
	sr.Close()
End Using
```

### StreamReader with encoding and "ToLower" added takes 8:39
```
Using sr As New StreamReader(filename, Encoding.Default)
	Dim aline As String
	While Not sr.EndOfStream
		aline = sr.ReadLine.ToLower
		If aline.IndexOf(srchText) <> -1 Then
			Console.WriteLine(aline)
		End If
	End While
	sr.Close()
End Using
```
### StreamReader with no encoding and no processing (obviously useless but thought it might be interesting) takes 2:56
```
Using sr As New StreamReader(filename, Encoding.Default)
	Dim aline As String
	While Not sr.EndOfStream
		aline = sr.ReadLine
	End While
	sr.Close()
End Using
```

**Comment by doug24** [comment]
Thanks, this is helpful.  A couple comments on your tests:

When you don't specify an encoding, StreamReader defaults to UTF8.  So in the first two tests, you may be seeing the difference between the correct encoding and an incorrect encoding.  On my system Encoding.Default is iso-8859-1. 
 
Doing a case sensitive search is faster in dnGrep, like you see in your tests.  dnGrep is using the ignore case version of string IndexOf rather than ToLower.  It looks like which is faster depends on the contents of the file.

dnGrep is using the StreamReader as you did in your tests, but the big problem in dnGrep right now is that it is not using the standard ReadLine method.  It's doing special processing to preserve the EOL characters, and that is killing the performance.  The code has been this way for 5 years now, and I know the EOL is needed in at least one case, so I'll have to do more analysis and testing to see if there are others.  It may be that they can be dropped in most cases.  Also for Search, I think it may not be necessary to preserve the actual EOL characters, but just use a unix newline all the time - so it can be added on to the line only when needed.  For Replace, I recently added code to preserve the original file's EOL characters, and it gets the EOL from the first line of the file.

I think we can improve search performance significantly. I'm testing with a 1.9GB text file with 1,000,000 lines.   
Case Sensitive - before: 1:02  after:  0:24
Case Insensitive - before: 1:10  after: 0:29
where "before" is the current release, and "after" is dropping the EOL characters.

**Comment by doug24** [comment]
This looks like it's going to be a lot harder to fix than I thought.  There is a *lot* of code that depends on preserving the EOL characters.  For example, they are needed for the index within the file for the results pane and file preview.

**Comment by captkronos** [comment]
Thanks for looking into this. Two suggestions:

1. Offer the user a choice when undertaking a search of either (a) being able to utilise all the current dnGrep features and have a full output of the results, or (b) being constrained by limited search functionality and limited output of results, but with the benefit of being faster.

2. Since it seems that after five years I am the first to raise this performance issue, assume that the current version of dnGrep meets the vast majority of user needs, and keep dnGrep as it is.

**Comment by doug24** [comment]
@captkronos have you had a chance to try the 2.9.49 or 2.9.53 build at https://github.com/dnGrep/dnGrep/releases ?  I would expect it to be about as fast as the test cases you posted above.

**Comment by captkronos** [comment]
Sorry, I didn't notice there was an update. I have just tried 2.9.49 and you have indeed managed to deliver a very significant increase in performance. The search that previously took 19mins now takes just 7.5mins.

Great job and thanks.

**Comment by JVimes** [comment]
Thanks, Doug! Nice work. Sorry, I forgot to finish publishing 2.9.53. It'll be up in a minute.

**Comment by doug24** [comment]
Fixed in release 2.9.49 and later
